csharp/src/iceboxnet/ServiceManager.cs

```diff
@@ -441,295 +405,463 @@ async Task StartedAsync(IServiceObserverPrx observer, string[] services)
 
-                    if (addFacets)
+            /// <summary>Return true when the service is started, and returns false when the service is stopped. Other
+            /// states are not visible to the caller.
+            /// IMPORTANT: This property must be called with ServiceManager's _mutex locked!</summary>
+            internal bool IsStarted
+            {
+                get
+                {
+                    if (_pendingTask is Task pendingTask)
                     {
-                        // Add all facets created on the service communicator to the IceBox communicator
-                        // but renamed IceBox.Service.<service>.<facet-name>, except for the Process facet
-                        // which is never added
-                        foreach (KeyValuePair<string, IObject> p in communicator.FindAllAdminFacets())
+                        // We clear _pendingTask if it's completed; when get returns, _pendingTask is null or still
+                        // running.
+
+                        if (pendingTask.IsCompletedSuccessfully)
                         {
-                            if (!p.Key.Equals("Process"))
-                            {
-                                _communicator.AddAdminFacet(serviceFacetNamePrefix + p.Key, p.Value);
-                            }
+                            _started = !_started;
+                            _pendingTask = null;
+                        }
+                        else if (_pendingTask.IsFaulted || _pendingTask.IsCanceled)
+                        {
+                            _pendingTask = null; // _started remains the same
                         }
+                        // else _pendingTask is still running and we're still in the old state
                     }
+                    return _started;
                 }
+            }
 
-                try
+            internal string Name { get; }
+
+            private readonly string[] _args;
+
+            // A task that is changing _started to its opposite value.
+            private Task? _pendingTask;
+
+            private readonly IService _service;
+
+            private bool _started;
+
+            internal ServiceInfo(string name, IService service, Communicator? communicator, string[] args)
+            {
+                _args = args;
+                Communicator = communicator;
+                Name = name;
+                _service = service;
+                _started = true;
+            }
+
+            /// <summary>Starts this service.</summary>
+            /// <param name="serviceManager">The service manager.</param>
+            /// <param name="cancel">The cancellation token.</param>
+            /// <return>A task that completes successfully when the service is fully started.</return>
+            internal async Task StartAsync(ServiceManager serviceManager, CancellationToken cancel)
+            {
+                Task? startTask = null;
+
+                do
```

Comment:
We have pretty much the same code in StartAsync/StopAsync, could we have a single method that performs the start or stop depending on the state? Something along these lines:
```
            internal async Task WaitForPendingAndPerformAsync(
                ServiceManager serviceManager,
                bool start,
                Func<Task> perform,
                CancellationToken cancel)
            {
                Task? performTask = null;

                do
                {
                    // We may need to wait until the pending stop task completes
                    Task? waitTask = null;

                    lock (serviceManager._mutex)
                    {
                        if (IsStarted == start)
                        {
                            if (_pendingTask is Task pendingTask)
                            {
                                // wait for the state change to occur
                                waitTask = pendingTask;
                            }
                            else
                            {
                                // already in the correct state
                                return;
                            }
                        }
                        else
                        {
                            // If there's a pending task, we piggy-back on this task that will switch the state to
                            // desired state, else we create a new task to switch to the given state.
                            _pendingTask ??= perform();
                            performTask = _pendingTask;
                        }
                    }

                    if (waitTask != null)
                    {
                        try
                        {
                            await waitTask.WaitAsync(cancel);
                        }
                        catch
                        {
                            // If the await failed because this cancel was canceled, we want to throw, else we ignore
                            // the exception and loop back.
                            cancel.ThrowIfCancellationRequested();
                        }
                    }
                }
                while (performTask == null);

                await performTask.WaitAsync(cancel);
            }
```

Reply:
I think it's simpler to keep it as-is.